incubator/validator-json/src/main/java/io/aklivity/zilla/runtime/validator/json/JsonValueValidator.java

```diff
@@ -66,54 +68,30 @@ public JsonValidator(
         }).collect(Collectors.toList());
         this.catalog = catalogs.get(0).schemas.size() != 0 ? catalogs.get(0).schemas.get(0) : null;
         this.handler = handlersById.get(catalogs.get(0).id);
+        this.subject = catalog != null &&
+            catalog.subject != null ?
+            catalog.subject : config.subject;
     }
 
-    @Override
-    public boolean read(
-        DirectBuffer data,
-        int index,
-        int length)
-    {
-        return validate(data, index, length);
-    }
-
-    @Override
-    public boolean write(
-        DirectBuffer data,
-        int index,
-        int length)
-    {
-        return validate(data, index, length);
-    }
-
-    private boolean validate(
-        DirectBuffer data,
-        int index,
-        int length)
+    String fetchSchema(
+        int schemaId)
     {
         String schema = null;
-        int schemaId = catalog != null ? catalog.id : 0;
-
-        byte[] payloadBytes = new byte[length];
-        data.getBytes(0, payloadBytes);
-
         if (schemaId > 0)
         {
             schema = handler.resolve(schemaId);
         }
         else if (catalog != null)
         {
-            schemaId = handler.resolve(catalog.subject, catalog.version);
-            if (schemaId != 0)
+            if (schemaId > 0)
             {
                 schema = handler.resolve(schemaId);
             }
         }
-
-        return schema != null && validate(schema, payloadBytes);
+        return schema;
     }
 
-    private boolean validate(
+    boolean validate(
```

Comment:
A lot of work is done in this method on every invocation, even though much of it is guaranteed to be identical if the schemaId is constant.

Lets aim to optimize this a bit more by hoisting more of the predictable work to be done just once if possible.